File: api/src/main/java/in/kyle/mcspring/command/CommandScanner.java
package in.kyle.mcspring.command;

import org.bukkit.plugin.Plugin;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.util.Map;
import java.util.Set;

import in.kyle.mcspring.util.SpringScanner;
import lombok.AllArgsConstructor;
import lombok.var;

@Component
@AllArgsConstructor
class CommandScanner implements ApplicationContextAware {
    
    private final CommandController controller;
    private final SpringScanner scanner;
    private final CommandFactory commandFactory;
    
    @Override
    public void setApplicationContext(ApplicationContext ctx) throws BeansException {
        Map<Method, Object> scan = scanner.scanMethods(Command.class);
        for (var e : scan.entrySet()) {
            Command command = e.getKey().getAnnotation(Command.class);
            String name = command.value();
            var bukkitCommand = commandFactory.makeCommand(e.getKey(), e.getValue(), name);
            controller.registerCommand(bukkitCommand);
        }
    }
    
    interface CommandFactory {
        org.bukkit.command.Command makeCommand(Method method, Object object, String name);
    }
}

File: mcspring-api/mcspring-base/src/main/kotlin/in/kyle/mcspring/command/registration/CommandScanner.kt
package `in`.kyle.mcspring.command.registration

import `in`.kyle.mcspring.command.Command
import `in`.kyle.mcspring.util.SpringScanner
import org.springframework.context.ApplicationContext
import org.springframework.context.ApplicationContextAware
import org.springframework.stereotype.Component
import java.lang.reflect.Method

@Component
internal class CommandScanner(
        private val scanner: SpringScanner,
        private val commandRegistration: CommandRegistration
) : ApplicationContextAware {

    private val registeredCommands: MutableSet<Method> = mutableSetOf()

    override fun setApplicationContext(ctx: ApplicationContext) {
        val scan = scanner.scanMethods(Command::class.java)
        scan.filterKeys { it !in registeredCommands }
                .filterKeys { it.returnType in arrayOf(Void::class.java, String::class.java) }
                .forEach { (key, value) ->
                    val command = key.getAnnotation(Command::class.java)
                    commandRegistration.register(command, key, value)
                    registeredCommands.add(key)
                }
    }
}
